File: python3/ValidParantheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        
        memo = {"]" : "[", ")" : "(", "}" : "{"}
        for char in s:
            if char in "({[":
                stack.append(char)
            else:
                if len(stack) > 0 and memo[char] == stack[-1]:
                    stack.pop()
                else:
                    return False
        
        return len(stack) == 0

File: c++/ValidParantheses.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> close;
        
        close[')'] = '(';
        close['}'] = '{';
        close[']'] = '[';
        
        stack<char> st;
        set<char> open = {'(', '{', '['};
        for (char item : s) {
            if (open.find(item) != open.end()) {
                st.push(item);
            } else {
                if (!st.size() || close[item] != st.top()) {
                    return false;
                }
                st.pop();
            }
        }
        
        if (st.size())
            return false;
        
        return true;
    }
};
